cairo: 1.16.0 -> 1.18.0

alyssais opened the pull request:
## Hydra comparisons

- [1800235](https://hydra.nixos.org/eval/1800235?compare=1799588&full=0) (complete)

## Description of changes

 - libiconv and expat are no longer used.
 - libintl and xcbutil no longer appear to be used.
 - The GL backend has been removed.
 - We can no longer disable PDF support, since something always propagates zlib.
 - Not sure whether the Requires.private is still needed, so I've left it.

Fixes #256519.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

bobby285271 committed:
gst_all_1.gst-plugins-good: Add missing libGL

With cairo 1.18.0 bump, cairo-gl will be dropped, and cairo will no
longer propagate libGL, this leads to: Run-time dependency
gstreamer-gl-prototypes-1.0 found: NO

glesv2 is in gstreamer-gl-prototypes-1.0's Requires, so adding libGL.

bobby285271 committed:
fcitx5: Add missing expat and xcbutil

These are direct dependencies for fcitx5, as listed in CMakeLists.txt.

When cairo is updated to 1.18.0, it will no longer propagates both.

alyssais committed:
cairo: 1.16.0 -> 1.18.0

 - libiconv and expat are no longer used.
 - libintl and xcbutil no longer appear to be used.
 - The GL backend has been removed.
 - We can no longer disable PDF support, since something always
   propagates zlib.
 - Not sure whether the Requires.private is still needed, so I've left
   it.

RaitoBezarius commented:
Do you want to get a Hydra jobset for this?

bobby285271 commented:
The only build fixes for this in GNOME 45 PR so far are 2968007ad4786fd9a3ea731a7b91f31ac5f282cb and cd6c696b0dbc03961c8f6127c767c18007803fb8 if you wish to do this in separate PR (but the GNOME 45 does need the bump for librsvg, glycin-loaders, loupe, etc)

alyssais commented:
@vcunat could we get a Hydra jobset?

bobby285271 commented:
(The GNOME 45 PR now also have 517d24e6accda409b0786d317619027e022d1354, ed5e1e18cc13416e96244a6978277a3a43684459, and I-don't-know-why 12b140ca8d9bd7762c3e2f86a608a42a83f3ae51)

vcunat commented:
Jobset created: https://hydra.nixos.org/jobset/nixpkgs/pr-257301-cairo-1.18

I doubt that it will be possible to get significant amounts of builds for this PR done this week, with all-rebuilding rush starting on Wednesday due to security fixes.  I guess we'll also wait with gnome jobset until this progresses?  (it also contains the update and it doesn't make much sense to me for these jobsets to compete against each other for builders)

vcunat commented:
I triggered it for x86_64-linux and aarch64-darwin for now.

bobby285271 committed:
weston: Add missing libGL

Cairo will soon no longer propagate libGL.

Run-time dependency egl found: NO (tried pkgconfig and cmake)

bobby285271 committed:
string-machine: Add missing libGL

Cairo will soon no longer propagate libGL. The DPF submodule needs this:

src/pugl/pugl.h:39:14: fatal error: GL/gl.h: No such file or directory
   39 | #    include <GL/gl.h>
      |              ^~~~~~~~~
compilation terminated.

bobby285271 committed:
stone-phaser: Add missing libGL

Cairo will soon no longer propagate libGL. The DPF submodule needs this:

src/pugl/pugl.h:39:14: fatal error: GL/gl.h: No such file or directory
   39 | #    include <GL/gl.h>
      |              ^~~~~~~~~
compilation terminated.

(See also the fix for string-machine, which has the same build issue)

bobby285271 committed:
mission-center: Add missing libGL

Cairo will soon no longer propagate libGL.

ld: cannot find -lEGL: No such file or directory

bobby285271 committed:
hyprpicker: Add missing libGL

Cairo will soon no longer propagate libGL.

/build/source/src/debug/../includes.hpp:30:10: fatal error: GLES3/gl32.h: No such file or directory
   30 | #include <GLES3/gl32.h>
      |          ^~~~~~~~~~~~~~
compilation terminated.

bobby285271 committed:
hyprpaper: Add missing libGL

Cairo will soon no longer propagate libGL.

/build/source/src/events/../includes.hpp:31:10: fatal error: GLES3/gl32.h: No such file or directory
   31 | #include <GLES3/gl32.h>
      |          ^~~~~~~~~~~~~~
compilation terminated.

bobby285271 committed:
fbida: Add missing libGL

Cairo will soon no longer propagate libGL.

Build dependencies missing for fbi and/or fbpdf.
  fbi   needs:  freetype2 fontconfig libdrm libexif libpng libtiff-4 pixman-1
  fbpdf needs:  libdrm poppler-glib gbm egl epoxy pixman-1

While at it also explicitly add pixman.

vcunat commented:
Builds finished, FYI.

alyssais committed:
cogl: force linking against libGL if used

Otheriwse, if Cairo's GL backend is disabled, libGL doesn't end up in
cogl's rpath, so applications using cogl (like gthumb) will fail to
launch when they search for libGL and can't find it.

vcunat commented:
Uh, the push... do we really need to rebuild 100k builds now?

alyssais commented:
No, go ahead and cancel it.

vcunat commented:
I had one eval per day timer in there, but I triggered this one myself :-)  I thought it was mainly the two extra commits shown by GitHub UI.

vcunat commented:
So I suppose you'll soon consider this ready to merge, without another Hydra eval?  There are only 10 days left to merge breaking changes that should reach NixOS 23.11.

vcunat commented:
OK, I stopped the eval timer.  Ping me when you want it.

alyssais committed:
rubyPackages.cairo-gobject: add expat build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
wxSVG: add expat build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
wmfocus: add expat build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
kapitonov-plugins-pack: add xcbutil build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
i3lock-color: add xcbutil build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
i3lock: add xcbutil build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
industrializer: add libGL build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
clapper: add libGL build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
hyprland: add libGL build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
i3lock-blur: add libGL build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
jay: add libGL build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
sway: add libGL build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
wayfire: add libGL build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
celluloid: add libGL build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
squeak: add libGL build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais committed:
gwyddion: add libGL build input

Currently, this is propagated from cairo, but that will soon no longer
be the case.

alyssais commented:
@bobby285271 I've fixed cogl in a different way, just linking libGL and avoiding the need for a patch.

alyssais commented:
@vcunat I've now addressed all the direct new build failures from the last Hydra run, so I think it's time for another.

fabianhjr reviewed (approved):
Been using cairo 1.18.0 as part of GNOME 45 branch: https://github.com/NixOS/nixpkgs/pull/247766/files#diff-c1b6675856a14491b8c610758d37ad4372e04a3f25a6049926c738f21121e5b0

Eval looks great with minimal breakage ( https://hydra.nixos.org/eval/1800745 )

vcunat commented:
After taking various things into consideration, let's finish this during the upcoming `staging-next`.

bobby285271 commented:
And we probably still need to fix cross there I guess (did not test, but I saw the `ipc_rmid_deferred_release` lines commented) :thinking:

alyssais commented:
Oh, whoops.

trofi commented:
I noticed that `cairo` bump changed `x11Support` condition from `!stdenv.isDarwin` to `!stdenv.isDarwin || true` which seems to be an effective `true`. Should it be just `true`? Or it's a debug leftover that should be `!stdenv.isDarwin`?

fabianhjr commented on pkgs/development/libraries/cairo/default.nix:
```diff
@@ -1,12 +1,8 @@
-, x11Support? !stdenv.isDarwin, libXext, libXrender
+{ lib, stdenv, fetchurl, fetchpatch, gtk-doc, meson, ninja, pkg-config, python3
+, docbook_xsl, fontconfig, freetype, libpng, pixman, zlib
+, x11Support? !stdenv.isDarwin || true, libXext, libXrender
```
Good observation, my instinct thinks this is an unintended change and `|| true` should be removed but tagging @NixOS/darwin-core for decision on disabling or enabling on darwin by default.

fabianhjr commented:
Mhhh, and this is a year old. Probably best to leave enabled then? (It has been part of the last 2 stable releases)

reckenrode commented:
I would prefer not to pull in X11 libraries unnecessarily, but what’s the effect on closure size?
